<a href="https://colab.research.google.com/github/Ila021/MCIO1/blob/main/ModeloPL21162.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto realizado por Ignacio Lara Alba 21162

# Resolución de modelos de optimización con Python

En primer lugar, la ventaja principal de usar Python es la capacidad de resolver modelos de manera rápida sin tener que depender de softwares no Open Sources teniendo que hacer la inversión en un software externo. Ademá, el uso de Python permite comprender lo que se realiza en cada línea de código si tener que comprender la interfaz gráfica de un programa externo.

En adición a esto, Python dispone de infinidad de recursos gratuitos disponibles para cualquier usuario y pudiendo implementar otras herramientas de regresión lineal e inteligencia artificial que nos permitan complementar nuestros modelos de programación lineal haciendolos más robustos ante cualquier situación y también la implementación de interfaces gráficas para que se pueda implementar el programa como un ejecutable .exe (al igual que AIIMS o Excel).

# Exposición de las herramientas utilizadas

Ante la infinidad de librerías de Python que permiten la resolución de modelos de programación lineal, he decidido utilizar Pulp ya que a diferecia de otras liberías como Gekko, Pulp permite el cálculo de los criterios del Simplex, precios sombras y el valor que toman las holguras, si estas formasen parte de las variables básicas de la solución óptima del modelo.

También, empleo la liberia Panda, para mostrar los valores de mis precios sombras y holguras en una tabla ordenada, ya que python me la devuelve como una Array poco legible.

Y por último, destaco el uso de Jupyter y Google Colabs, ya que son herramientas sencillas que permiten la escritura de documentos en Markdown (permitiendo su reescritura en Latex) para explicar de manera amigable el código y la capacidad de realizar proyectos colaborativos de manera online.   


In [ ]:
#!pip install pulp (instalación de las liberías antes de escibir el modelo)
#!pip install pandas

## Importación de las librerías y simplificación de sus nombres para su uso de manera sencilla en el código

In [ ]:
import pulp as p
import pandas as pd

## Definición del modelo de Programación Lineal

A partir del siguiente modelo de una fábrica de vehículos de combustión:

$$
max: z=220\cdot x_1+350\cdot x_2
$$
$$
x_1+x_2\geq50
$$
$$
0.4x_1+0.75x_2\leq100
$$
$$
x1,x2\geq0
$$

Donde $x_1$ es el número de producción de coches y $x_2$ es le número de producción de camiones.

Se desea maximizar el beneficio de la planta sabiendo el beneficio unitario de coches y camiones.

Y existe una limitación de piezas y una exigencia comercial de 50 unidades.

## Definición de un problema de programación lineal de máximos

In [ ]:
modelo = p.LpProblem('Producción', p.LpMaximize) #Definición del modelo

## Definición de las variables

Donde lowBound es la restricción de no negatividad y cat es la categoria de la varible donde se le indica que puede tomar un valor float (con decimales). Muy importante definirlo así para evitar errores en la resolución del modelo.

Si fuera binario o entero pondríamos cat='p.LpBinary' o cat='p.LpInteger'

In [ ]:
x1= p.LpVariable('Coches', lowBound=0, cat= p.LpContinuous)
x2= p.LpVariable('Camiones', lowBound=0, cat= p.LpContinuous)

## Definición de la función objetivo

In [ ]:
modelo += 220*x1 + 350*x2, 'Funcion objetivo'

## Definición de las restricciones
 Como se puede apreciar la sintaxis de las restricciones es muy sencilla

In [ ]:
modelo += x1 + x2 >= 50
modelo += 0.4*x1+ 0.75*x2 <=100

## Comprobación de que la ejecución correcta

In [ ]:
modelo

Producción:
MAXIMIZE
350*Camiones + 220*Coches + 0
SUBJECT TO
_C1: Camiones + Coches >= 50

_C2: 0.75 Camiones + 0.4 Coches <= 100

VARIABLES
Camiones Continuous
Coches Continuous

## Resolución del modelo a apartir del método del Simplex

Devuelve un 1 si la resolución ha sido correcta. Se podría añadir un if else  para mostrar o no los resultados finales.

In [ ]:
modelo.solve()

1

## Exposición de los resultados finales

In [ ]:
print("Número de coches", x1.varValue)
print("Número de caminones", x2.varValue)
print("Beneficio", p.value(modelo.objective))

Número de coches 250.0
Número de caminones 0.0
Beneficio 55000.0


## Análisis de sensibilidad

Se muestran gracias al paquete Panda, en una tabla el valor de los precios sombras y el valor de las holguras. Se emplea un bucle for para la conversión de la array que devuelve Pulp para construir la tabla de manera correcta.

In [ ]:
sensibilidad =[{'Restricción':i, 'Precio Sombra':j.pi, 'Holgura': j.slack}
               for i, j in modelo.constraints.items()]
print(pd.DataFrame(sensibilidad))

  Restricción  Precio Sombra  Holgura
0         _C1           -0.0   -200.0
1         _C2          550.0     -0.0


Se observa que se fabrican solo coches en 250 unidades y sobran 200 unidades en el compromiso comercial. Se emplean todos los recursos disponibles.

© Modelo y código realizado por Ignacio Lara Alba para la asignatura de MCIO 1 del grado de GITI de la UPM 20/03/2024.